### PR TITLE
Uplift `FnSig` to `rustc_type_ir`

### PR DESCRIPTION
**Pull request by compiler-errors**
:3 

r? lcnr

### CODE REVIEW

**Comment by compiler-errors** [comment]
@bors try @rust-timer queue

Removed some `&self` in favor of `self` so let's perf this too

File: compiler/rustc_type_ir/src/inherent.rs [hunk 1]
```diff
@@ -26,6 +26,21 @@ pub trait Ty<I: Interner<Ty = Self>>:
...
 }
 
+pub trait Tys<I: Interner<Tys = Self>>:
+    Copy + Debug + Hash + Eq + IntoIterator<Item = I::Ty> + Deref<Target: Deref<Target = [I::Ty]>>
+{
+    fn split_inputs_and_output(self) -> (I::FnInputTys, I::Ty);
+}
+
+pub trait Abi<I: Interner<Abi = Self>>: Copy + Debug + Hash + Eq {
+    // is `extern "Rust"`?
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/inherent.rs:36]:
```suggestion
    /// Whether this ABI is `extern "Rust"`.
```

**Comment by lcnr** [comment]
r=me after nit + CI

**Comment by compiler-errors** [comment]
I feel like this branch is beyond repair

**Pull request closed without merging**